Каркас приложения

Natal1a-Chuklina opened the pull request.

Natal1a-Chuklina committed:
initial spring commit

Natal1a-Chuklina committed:
[+] FilmController.class - добавлен контроллер для обработки запросов, связанных с фильмами
[+] UserController.class - добавлен контроллер для обработки запросов, связанных с пользователями
[+] AlreadyExistException.class - добавлено исключение для ситуаций, когда происходит попытка создать уже существующую сущность
[+] NotFoundException.class - добавлено исключение для ситуаций, когда сущность не найдена
[+] StartDateValidation.annotation - добавлена аннотация для ограничения значения поля с типом LocalDate
[+] StartDateValidator.class - добавлен валидатор для аннотации StartDateValidation
[+] Film.class - добавлена модель фильма
[+] User.class - добавлена модель пользователя
[+] FilmorateApplication.class - добавлен стартовый класс
[+] StartDateValidatorTest - добавлены тесты корректности работы кастомного валидатора
[+] FilmorateApplicationTests - добавлен тест запуска приложения

z8q commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,53 @@
+@RestController
+@RequestMapping("/films")
+public class FilmController {
+    private static final Logger logger = LoggerFactory.getLogger(FilmController.class);
```
Можно вместо этой строки попробовать поставить над классом `@Slf4j`
Вызов логгера будет такого вида:
log.info("Some info");

Natal1a-Chuklina replied:
Сделала.

z8q commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,53 @@
+    @PostMapping
+    public Film create(@Valid @RequestBody Film film) {
+        if (films.containsKey(film.getId())) {
+            logger.info("Выполнена попытка добавить фильме с уже существующим id = {}.", film.getId());
```
При ошибках стоит использовать log.error

Natal1a-Chuklina replied:
Уровень error приберегла для более тяжелых ситуаций. :) Переделала на уровень warn.

z8q commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,53 @@
+            throw new NotFoundException(String.format("Фильм с идентификатором %d не найден", film.getId()));
+        }
+
+        BeanUtils.copyProperties(film, films.get(film.getId()), "id");
```
Чем такой способ лучше `map.replace(key, value);`?

Natal1a-Chuklina replied:
Только тем, что можно задать поля, которые при копировании будут игнорироваться и останутся прежними.

z8q commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -0,0 +1,53 @@
+        return users.values();
+    }
+
+    @PostMapping
```
Будет правильно добавлять значения к аннотациям контроллеров, например:
@PostMapping("/create")
Если будет два POST метода, все равно это придется сделать

Natal1a-Chuklina replied:
Я бы с радостью сама что-то поменяла, но в ТЗ нам дали тесты в postman (https://github.com/yandex-praktikum/java-filmorate/blob/controllers-films-users/postman/sprint.json), по которым можно судить о корректности работы приложения, как я понимаю, они так же запускаются при пул реквесте. Поэтому возможности что-то сильно менять и как-то экспериментировать у меня нет.

z8q replied:
Хорошо,
Но если возьмем этот пример, что мы используем `@RequestMapping("/users")`, а на самих методах только аннотации без определения.
Создадим два post метода, 
При post-вызове /users будет ошибка

z8q commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -0,0 +1,36 @@
+@ToString
+@Setter
+public class Film {
+    @EqualsAndHashCode.Include
```
Мне не очень нравится такая идея.
Если где-то в новых сервисных классах будем создавать объекты Film:
Film x = new Film(name1, description1, LocalDate.of(2023, 2, 1), 1);
Film y = new Film(name2, description2, LocalDate.of(2023, 2, 2), 2);
что даст x.equals(y)?
Можно посоветоваться с наставником

Natal1a-Chuklina replied:
Переделала.

z8q commented on src/test/java/ru/yandex/practicum/filmorate/model/validate/StartDateValidatorTest.java:
```diff
@@ -0,0 +1,34 @@
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+class StartDateValidatorTest {
```
Можно было бы вызовы контроллеров тоже потестировать

Natal1a-Chuklina replied:
Предполагаю, что в следующих модулях нам подробно расскажут про тестирование и я протестирую контроллеры. Пока просили протестировать только валидацию.

z8q commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -0,0 +1,36 @@
+    private String name;
+    @Size(max = 200)
+    private String description;
+    @StartDateValidation(date = "28.12.1895", message = "Дата релиза должна быть позже 28 декабря 1895 года.")
```
Тут интересно придумано :)

z8q commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -0,0 +1,36 @@
+    public User(String email, String login, String name, LocalDate birthday) {
+        this.email = email;
+        this.login = login;
+        this.name = (name == null || name.isBlank()) ? login : name;
```
В конструкторе проверку имени не сразу заметил.
Возможно было бы лучше выводить проверку и setName if blank в какой-то метод

Natal1a-Chuklina replied:
Переделала

z8q reviewed (changes requested):
Привет, Наталья!

Оставил комментарии :)

Natal1a-Chuklina committed:
[r] FilmController.class - создание логгера заменено на аннотацию @Slf4j
[r] UserController.class - создание логгера заменено на аннотацию @Slf4j
[+] Film.class - в методы equals и hashCode добавлены все поля класса
[+] User.class - в методы equals и hashCode добавлены все поля класса
[r] User.class - инициализация поля name вынесена из конструктора в отдельный метод

z8q commented on src/main/java/ru/yandex/practicum/filmorate/FilmorateApplication.java:
```diff
@@ -0,0 +1,16 @@
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+
+/* Привет, Кирилл!
+Я не стала делать абстрактный контроллер и наследовать от него два других, мне показалось, что это не сильно хорошая идея.
```
Согласен. Не уверен, что это может пригодится. 
Общую логику можно потом будет реализовать в сервисных классах.
Если только через абстрактный класс определять, какие методы должны обязательно реализовывать контроллеры. Но не знаю, есть ли смысл...)
